### PR TITLE
[framework] Animatable.fromCallback and code snippet

### PR DESCRIPTION
**Pull request by jonahwilliams**
Add an easy way to transform an animation using a callback

### CODE REVIEW

File: packages/flutter/lib/src/animation/animation.dart [hunk 1]
```diff
@@ -298,3 +328,30 @@ class _ValueListenableDelegateAnimation<T> extends Animation<T> {
   @override
   T get value => _transformer?.call(_listenable.value) ?? _listenable.value;
 }
+
+// An implementation of an animation that delegates to a parent animation and applies a
+// potentially type-changing transform.
+class _MappedAnimation<S, T> extends Animation<T> {
```
Comment (goderbauer) [inline packages/flutter/lib/src/animation/animation.dart:334]: Could use `AnimationWithParentMixin`?

File: packages/flutter/lib/src/animation/animation.dart [hunk 2]
```diff
@@ -230,6 +233,32 @@ abstract class Animation<T> extends Listenable implements ValueListenable<T> {
     return child.animate(this as Animation<double>);
   }
 
+  /// Create a new [Animation] that transforms this animation using the provided
```
Comment (goderbauer) [inline packages/flutter/lib/src/animation/animation.dart:236]: In theory, you can achieve the same thing by implementing a custom `Animatable` class (only have to implement the `transform` method with the closure you would provide here) and then use `Animation.drive`.

This API avoids the extra class definition, I guess. Is that really such a hassle?

Comment (goderbauer) [reply packages/flutter/lib/src/animation/animation.dart:236]: Oh, and I guess Animation.drive only works if the parent is a `Animation<double>`. This one in theory allows any Animation to be transformed into any other... Is that a feature you need?

Comment (jonahwilliams) [reply packages/flutter/lib/src/animation/animation.dart:236]: Ahh maybe, let me look into that more.

**Comment by jonahwilliams** [comment]
Drive solves this problem. Perhaps I should re-tool this PR into an additional of how to use Animation.drive?

**Comment by goderbauer** [comment]
> Drive solves this problem. Perhaps I should re-tool this PR into an additional of how to use Animation.drive?

That would be much appreciated!

File: packages/flutter/lib/src/animation/tween.dart [hunk 3]
```diff
@@ -29,6 +32,14 @@ abstract class Animatable<T> {
   /// const constructors so that they can be used in const expressions.
   const Animatable();
 
+  /// Create a new [Animatable] from the provided [callback].
+  ///
+  /// See also:
+  ///
+  ///  * [Animation.drive], which provides an example for how this can be
+  ///    used.
+  const factory Animatable.fromCallback(AnimatableCallback<T> callback) = _CallbackAnimatable<T>;
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/animation/tween.dart:41]: Just a _little_ bit more API. It felt a bit silly to create a class just to wrap a callback in the sample code

File: packages/flutter/lib/src/animation/animation.dart [hunk 4]
```diff
@@ -224,6 +241,8 @@ abstract class Animation<T> extends Listenable implements ValueListenable<T> {
   ///  * [CurvedAnimation], an alternative to [CurveTween] for applying easing
   ///    curves, which supports distinct curves in the forward direction and the
   ///    reverse direction.
+  ///  * [Animatable.fromCallback], which allows creating an animatable from an
```
Comment (goderbauer) [inline packages/flutter/lib/src/animation/animation.dart:244]: nit: animatable -> Animatable (or `[Animatable]`)?

File: packages/flutter/lib/src/animation/animation.dart [hunk 5]
```diff
@@ -216,6 +216,23 @@ abstract class Animation<T> extends Listenable implements ValueListenable<T> {
...
+  /// of red.
+  ///
+  /// ```dart
+  /// Animation<Color> _offset1 = Animation<double>.fromValueListenable(_scrollPosition)
+  ///   .drive(Animatable<Color>.fromCallback((double value) {
+  ///     return Color.fromRGBO(value.round() % 255, 0, 0, 1);
+  ///   }));
+  /// ```
+  ///
+  /// {@end-tool}
```
Comment (goderbauer) [inline packages/flutter/lib/src/animation/animation.dart:235]: Nice doc!

File: packages/flutter/lib/src/animation/tween.dart [hunk 6]
```diff
@@ -17,6 +17,9 @@ export 'curves.dart' show Curve;
 // late Animation<Offset> _animation;
 // late AnimationController _controller;
 
+/// A typedef used to create an [Animatable] from a callback.
```
Comment (goderbauer) [inline packages/flutter/lib/src/animation/tween.dart:20]: nit: As per style guide link this doc directly to its use by including `[Animatable.fromCallback]` in the doc comment.

**Review by goderbauer — APPROVED** [review]
LGTM

**Pull request merged**